config/config_test.go

```diff
@@ -0,0 +1,235 @@
+	})
+
+	t.Run("valid: one env variable found within the same string", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "abc ${GORAY} abc")
+
+		assert.Equal(t, "abc some_value abc", got)
+		assert.Nil(t, err)
+	})
+
+	t.Run("valid: two env variables found within the same string", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${GORAY}${THREADS_COUNT}")
+
+		assert.Equal(t, "some_value5", got)
+		assert.Nil(t, err)
+	})
+
+	t.Run("valid: two env variables found within the same string", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${GORAY}${THREADS_COUNT}")
+
+		assert.Equal(t, "some_value5", got)
+		assert.Nil(t, err)
+	})
+
+	t.Run("valid: no env variables within the string", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${NOT AN ENV")
+
+		assert.Equal(t, "${NOT AN ENV", got)
+		assert.Nil(t, err)
+	})
+
+	t.Run("invalid: env variable not found", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${SOME_NAME}")
+
+		assert.Equal(t, "", got)
+		assert.Equal(t, "environment variables not found: SOME_NAME", err.Error())
+	})
+
+	t.Run("invalid: one env variable found, one not found", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${GORAY}${SOME_NAME}")
+
+		assert.Equal(t, "", got)
+		assert.Equal(t, "environment variables not found: SOME_NAME", err.Error())
+	})
+
+	t.Run("invalid: multiple env variables not found", func(t *testing.T) {
+		got, err := interpolate(testLookupEnv, "${SOME_NAME1} ${SOME_NAME2} ${SOME_NAME3}")
+
+		assert.Equal(t, "", got)
+		assert.Equal(t, "environment variables not found: SOME_NAME1, SOME_NAME2, SOME_NAME3", err.Error())
+	})
+}
+
+func withTestReadFile(rc *ReadConf) {
+	rc.readFile = testReadFile
+}
+
+func testReadFile(name string) ([]byte, error) {
+	return []byte(`
+		[workspace]
+		version = "1.0.193"
+		root= "."
+		policies="./policy"
+		ignore-list=[".git", ".docker"]
+		go-arch="HOME: ${GORAY}"
+		
+		[analysis]
+		KRAIL1001 = {level = "info", target= "./${GORAY}"}
+		
+		[analysis.SE]
+		level = "critical"
+		target = "./internal"
+	`), nil
+}
+
+func withTestReadFileErr(rc *ReadConf) {
+	rc.readFile = testReadFileErr
+}
+
+func testReadFileErr(name string) ([]byte, error) {
+	return nil, errors.New("read file error")
+}
+
+func withTestReadToml(rc *ReadConf) {
+	rc.readToml = testReadToml
+}
+
+func testReadToml(data string, v interface{}) (toml.MetaData, error) {
+	return toml.MetaData{}, nil
+}
+
+func withTestReadTomlErr(rc *ReadConf) {
+	rc.readToml = testReadTomlErr
+}
+
+func testReadTomlErr(data string, v interface{}) (toml.MetaData, error) {
+	return toml.MetaData{}, errors.New("read toml error")
+
+}
+
+func withTestInterpolate(rc *ReadConf) {
+	rc.interpolate = testInterpolate
+}
+
+func testInterpolate(lookup LookupFn, strWithEnvs string) (string, error) {
+	return "interpolated string", nil
+}
+
+func withTestInterpolateErr(rc *ReadConf) {
+	rc.interpolate = testInterpolateErr
```

Comment:
what the point of all this `with*` test functions? U are truing to test a test code )) 

if u whant to test a struct constractor u should test it with a constructor functions and check the results

Reply:
The point is to set the mock functions for the `ReadConf`
```golang
type ReadConf struct {
	lookup      func(key string) (string, bool)
	readFile    func(name string) ([]byte, error)
	readToml    func(data string, v interface{}) (toml.MetaData, error)
	interpolate func(lookup LookupFn, strWithEnvs string) (string, error)
}
```
 struct to test `func NewFromFile(options ...ConfReadFileOptFn) *ReadConf (...)` behavior with mocked functions. 

How should I otherwise test that all the functions from `ReadConf` are called in the `NewFromFile(...)` and the errors are processed?